#! /bin/sh -e

if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch $pdir -f --no-backup-if-mismatch -p1 < $0;;
    -unpatch) patch $pdir -f --no-backup-if-mismatch -R -p1 < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# DP: From: Topi Miettinen <Topi.Miettinen@nic.fi>
# DP: build bash without using guessable file names in /tmp.

From: Topi Miettinen <Topi.Miettinen@nic.fi>
Sender: tom@pluto.nic.fi
To: 36027@bugs.debian.org
Subject: Bug#36027: Suggested patch to fix #36027
Date: Tue, 14 Dec 1999 22:05:53 +0200

The following patch attempts to fix the security problem.

-Topi

diff -ru ./aclocal.m4.orig ./aclocal.m4
--- ./aclocal.m4.orig	Wed Dec 30 18:06:41 1998
+++ ./aclocal.m4	Tue Dec 14 21:59:06 1999
@@ -316,13 +316,19 @@
 main()
 {
 DIR *dir;
-int fd;
-unlink("/tmp/not_a_directory");
-fd = open("/tmp/not_a_directory", O_WRONLY|O_CREAT, 0666);
+int fd, err;
+
+err = mkdir("/tmp/bash-aclocal", 0700); 
+if (err == -1) {
+  perror("mkdir");
+  exit (1);
+}
+fd = open("/tmp/bash-aclocal/not_a_directory", O_WRONLY|O_CREAT, 0666);
 write(fd, "\n", 1);
 close(fd);
-dir = opendir("/tmp/not_a_directory");
-unlink("/tmp/not_a_directory");
+dir = opendir("/tmp/bash-aclocal/not_a_directory");
+unlink("/tmp/bash-aclocal/not_a_directory");
+rmdir("/tmp/bash-aclocal");
 exit (dir == 0);
 }], bash_cv_opendir_not_robust=yes,bash_cv_opendir_not_robust=no,
     [AC_MSG_WARN(cannot check opendir if cross compiling -- defaulting to no)
@@ -845,7 +851,7 @@
 /* Add more tests in here as appropriate. */
 main()
 {
-int fd;
+int fd, err;
 
 #if defined (HAVE_MKFIFO)
 exit (0);
@@ -859,11 +865,17 @@
 exit (1);
 #endif
 
-fd = mknod ("/tmp/sh-np-autoconf", 0666 | S_IFIFO, 0);
+err = mkdir ("/tmp/bash-aclocal", 0700); 
+if (err == -1) {
+  perror ("mkdir");
+  exit (1);
+}
+fd = mknod ("/tmp/bash-aclocal/sh-np-autoconf", 0666 | S_IFIFO, 0);
 if (fd == -1)
   exit (1);
 close(fd);
-unlink ("/tmp/sh-np-autoconf");
+unlink ("/tmp/bash-aclocal/sh-np-autoconf");
+rmdir ("/tmp/bash-aclocal");
 exit(0);
 }], bash_cv_sys_named_pipes=present, bash_cv_sys_named_pipes=missing,
     [AC_MSG_WARN(cannot check for named pipes if cross-compiling -- 
defaulting to missing)


diff -ru support/rlvers.sh.orig support/rlvers.sh
--- ./support/rlvers.sh.orig	Mon Dec  6 03:58:51 1999
+++ ./support/rlvers.sh	Tue Dec 14 22:32:33 1999
@@ -6,7 +6,7 @@
 
 PROGNAME=`basename $0`
 
-trap 'rm -f /tmp/rlvers /tmp/rlvers.?' 0 1 2 3 6 15
+trap 'rm -rf /tmp/rlvers' 0 1 2 3 6 15
 
 # defaults
 CC=cc
@@ -41,7 +41,8 @@
 	echo "${PROGNAME}: attempting program compilation"
 fi
 
-cat > /tmp/rlvers.c << EOF
+mkdir -m 0700 /tmp/rlvers
+cat > /tmp/rlvers/rlvers.c << EOF
 #include <stdio.h>
 extern char *rl_library_version;
 
@@ -52,9 +53,9 @@
 }
 EOF
 
-if eval ${CC} -L${RL_LIBDIR} -o /tmp/rlvers /tmp/rlvers.c -lreadline -lcurses;
+if eval ${CC} -L${RL_LIBDIR} -o /tmp/rlvers/rlvers /tmp/rlvers/rlvers.c -lreadline -lcurses;
 then
-	v=`/tmp/rlvers`
+	v=`/tmp/rlvers/rlvers`
 else
 	if [ -n "$verbose" ] ; then
 		echo "${PROGNAME}: compilation failed: status $?"


